### PR TITLE
Add exception explanation, fixes #87

### LINKED ISSUE #87: Exceptions
## Exceptions in promises

If I understand it well, in case of async requests the async request method should NEVER EVER throw an exception, but return and reject a promise.

This is currently not the case:
https://github.com/php-http/guzzle6-adapter/blob/master/src/Guzzle6Promise.php#L66
## SPL or not

Although we agreed that SPL exceptions can be used to determine exceptions not closely related to the domain, we might want to reconsider this.

The point of the common Exception interface is that every exception can be caught. On the other side: creating a custom version of every SPL exception which implements our interface seems to be dumb.

The ideal solution would be creating domain exceptions for each case we would use an SPL one.

For example in this case:
https://github.com/php-http/guzzle6-adapter/blob/master/src/Guzzle6Promise.php#L66

This decision should take effect in the Promise interface as well: currently it is restricted to our Exception interface.

**Update:** Using anonymus classes would be a good solution, but that is only available in PHP 7. Maybe we could place custom extensions in one of our utils repo.

### PR DESCRIPTION
**Pull request by sagikazarmark**
@dbu @joelwurtz please help me with the explanation. :smile:

### CODE REVIEW

**Comment by dbu** [comment]
looks good to me. for async, we could mention in the text part that exceptions that should be handled like network issues and server problems are available from the promise, as they might not be known yet when the sendAsyncRequest method returns.

**Comment by sagikazarmark** [comment]
Better now?

(Should be squashed)

**Comment by dbu** [comment]
looks good to me, yes!

**Comment by sagikazarmark** [comment]
@joelwurtz ping

File: src/HttpClient.php [hunk 1]
```diff
@@ -21,7 +21,8 @@
      *
      * @return ResponseInterface
      *
-     * @throws Exception
+     * @throws Exception  If an error happens during processing the request.
+     * @throws \Exception If processing the request is impossible (eg. bad configuration).
```
Comment (joelwurtz) [inline src/HttpClient.php]: Can we use this exception naming instead:

``` php
@throws \Http\Exception
@throws \Exception
```

Currently it's feel a bit weird and i'm afraid of misunderstanding from user, if it's read like that

**Comment by joelwurtz** [comment]
Aside from my comment, :+1: for me

File: src/HttpClient.php (hunk 1 above)
Comment (dbu) [inline src/HttpClient.php]: agreed for this situation. but in general we should keep using the `use` statement for phpdoc as well.

**Pull request merged**